### ISSUE TITLE
[Bug] copy_ub_to_gm 缺少 padding 配置导致非对齐 blockLen 数据损坏

### ISSUE THREAD

**Issue by 2Sheldonlaw** [tile-ai/tilelang-ascend#1682]
## Bug Description

`copy_ub_to_gm` 在 UB→GM 搬运时缺少 `DataCopyPadExtParams` padding 配置。当输出 buffer 最后一维元素数 × sizeof(T) 不是 32 字节倍数时（如 `(M, 1)` float32 → blockLen = 4 字节），`DataCopyPad` 无法正确处理非对齐的 blockLen，导致数据损坏。

这影响 `T.reduce_max/min/sum` 使用 `(M, 1)` keepdim 输出的场景，以及任何 `T.copy` 从 UB 到 GM 时最后一维不满足 32 字节对齐的场景。

`copy_gm_to_ub`（GM→UB）有完整的 padding 逻辑（`isPad` / `rightPadding` / `DataCopyPadExtParams`），但 `copy_ub_to_gm`（UB→GM）没有——两者不对称。

**所有仓库 examples 碰巧规避了此问题**——examples 中 reduce 的 `(M, 1)` keepdim 输出后续被 V-pipe 操作使用（如 `T.tile.broadcast`、`T.tile.add` 等），不是直接 `copy_ub_to_gm` 输出到 GM。5 个有精度验证的 example 全部通过。

## Environment

- NPU: Ascend910_9362 (A3, dav-c220)
- CPU Architecture: aarch64
- CANN: 9.0.0
- torch: 2.7.1+cpu
- torch-npu: 2.7.1.post4
- Python: 3.11.4
- TileLang-Ascend: `5149acdb40b086d7cd84e30821f5759fa8699a14`
- Installation: build from source

## Code to Reproduce

```bash
source set_env.sh
python repro_reduce_bug14.py
```

<details>
<summary>完整复现代码</summary>

```python
import torch
import tilelang
import tilelang.language as T

tilelang.cache.clear_cache()


@tilelang.jit(out_idx=[0], target="ascendc")
def reduce_keepdim(M: int = 16, N: int = 128, dtype: str = "float32"):
    @T.prim_func
    def main(B: T.Tensor((M, 1), dtype)):
        with T.Kernel(1, is_npu=True) as (cid, vid):
            a_ub = T.alloc_ub((M, N), dtype)
            b_ub = T.alloc_ub((M, 1), dtype)
            T.tile.arith_progression(a_ub, 0.0, 1.0, M * N)  # V-pipe fill
            T.reduce_max(a_ub, b_ub, dim=-1)
            T.copy(b_ub, B)  # copy_ub_to_gm<float, 1, 16> — blockLen=4, not aligned
    return main


M, N = 16, 128
kernel = reduce_keepdim(M, N)
b = kernel()
torch.npu.synchronize()

a = torch.arange(M * N, dtype=torch.float32).reshape(M, N)
ref = a.max(dim=-1, keepdim=True).values
match = torch.equal(b.cpu(), ref)
max_diff = (b.cpu() - ref).abs().max().item()

print(f"reduce_max to (M,1) keepdim — blockLen = 1 * 4 = 4 bytes (not 32-byte aligned):")
print(f"  {'PASS' if match else 'FAIL'} (max_diff={max_diff})")
```

</details>

### Reproduction results

```
reduce_max to (M,1) keepdim — blockLen = 1 * 4 = 4 bytes (not 32-byte aligned):
  FAIL (max_diff=896.0)
  b[:4]: [127.0, 1151.0, 71.0, 199.0]
  ref[:4]: [127.0, 255.0, 383.0, 511.0]

Generated C++ copy_ub_to_gm call:
  tl::ascend::copy_ub_to_gm<float, 1, 16>(B[0], b_ub[0], 1, 16, 1);
  blockLen = maskShapeN * sizeof(float) = 1 * 4 = 4 bytes
  4 % 32 != 0 → DataCopyPad without DataCopyPadExtParams corrupts data
```

复现脚本使用 V-pipe src（`arith_progression`）填充输入数据，排除了 broadcast issue #1641 中的 MTE2→V 数据可见性问题，纯粹测试 `copy_ub_to_gm` 的 padding 缺失。

## Error Message

无编译错误、无运行时错误——**静默失败**。编译通过，运行通过，但输出数据损坏。

## Expected Behavior

`copy_ub_to_gm` 应与 `copy_gm_to_ub` 一样，在 `blockLen` 不是 32 字节倍数时使用 `DataCopyPadExtParams` 配置 padding，确保非对齐数据正确搬运。

## Actual Behavior

`copy_ub_to_gm`（`common.h:251-260`）调用 `DataCopyPad` 时只有 3 参数，缺少 `DataCopyPadExtParams`：

```cpp
// copy_ub_to_gm — 无 padding 处理（BUG）
template <typename T, uint32_t srcN, uint32_t srcM = 1>
copy_ub_to_gm(GlobalTensor<T> dst, LocalTensor<T> src,
              uint32_t realdstN, uint32_t maskShapeM, uint32_t maskShapeN) {
    AscendC::DataCopyExtParams dataCopyParams(
        maskShapeM, maskShapeN * sizeof(T), ...);
    AscendC::DataCopyPad(dst, src, dataCopyParams);  // 3 参数，无 padParams
}
```

对比 `copy_gm_to_ub`（`common.h:215-249`）有完整 padding：

```cpp
// copy_gm_to_ub — 有 padding 处理（正确）
bool isPad = (maskShapeN * sizeof(T)) % 32 != 0;
uint8_t rightPadding = isPad ? 1 : 0;
AscendC::DataCopyPadExtParams<T> padParams(isPad, 0, rightPadding, padValue);
AscendC::DataCopyPad(dst, src, dataCopyParams, padParams);  // 4 参数，含 padParams
```

不同 shape 的 blockLen 和结果对比：

| 输出 shape | `copy_ub_to_gm` 模板 | blockLen | 32B 对齐 | 结果 |
|-----------|---------------------|----------|:---:|------|
| `(16, 1)` | `<float, 1, 16>` | 4 字节 | ❌ | FAIL |
| `(16,)` | `<float, 16>` | 64 字节 | ✅ | PASS |
| `(1, 128)` | `<float, 128, 1>` | 512 字节 | ✅ | PASS |

## Suggested Solution

给 `copy_ub_to_gm` 添加与 `copy_gm_to_ub` 对称的 padding 逻辑：

```cpp
template <typename T, uint32_t srcN, uint32_t srcM = 1>
CATLASS_DEVICE void
copy_ub_to_gm(GlobalTensor<T> dstTensor, LocalTensor<T> srcTensor,
              uint32_t realdstN = 1, uint32_t maskShapeM = srcM,
              uint32_t maskShapeN = srcN) {
    bool isPad = (maskShapeN * sizeof(T)) % 32 != 0;
    uint8_t rightPadding = isPad ? 1 : 0;

    AscendC::DataCopyExtParams dataCopyParams(
        maskShapeM, maskShapeN * sizeof(T),
        (srcN - maskShapeN) * sizeof(T) / 32,
        (realdstN - maskShapeN) * sizeof(T), 0);
    AscendC::DataCopyPadExtParams<T> padParams(isPad, 0, rightPadding, 0);
    AscendC::DataCopyPad(dstTensor, srcTensor, dataCopyParams, padParams);
}
```

风险：`copy_ub_to_gm` 被广泛调用，添加 padding 可能影响其他场景的 copy 行为。需要全量回归测试。

## Additional Context

**Backend:**
- [x] Ascend C backend
- [ ] PTO backend (pto 后端在 dav-c220 上完全不可用，无法测试)
- [ ] AscendNPU IR backend

**影响范围**：所有 2D 输出 buffer 最后一维元素数 × sizeof(T) 不是 32 字节倍数的 `T.copy` UB→GM 操作：
- float32：最后一维需要是 8 的倍数（`N * 4 % 32 == 0` → `N % 8 == 0`）
- float16：最后一维需要是 16 的倍数（`N * 2 % 32 == 0` → `N % 16 == 0`）

<details>
<summary>源码调用链</summary>

### TileLang C++ 模板

`src/tl_templates/ascend/common.h:251-260` — `copy_ub_to_gm`（无 padding）

```cpp
template <typename T, uint32_t srcN, uint32_t srcM = 1>
CATLASS_DEVICE void
copy_ub_to_gm(GlobalTensor<T> dstTensor, LocalTensor<T> srcTensor,
              uint32_t realdstN = 1, uint32_t maskShapeM = srcM,
              uint32_t maskShapeN = srcN) {
    AscendC::DataCopyExtParams dataCopyParams(
        maskShapeM, maskShapeN * sizeof(T),
        (srcN - maskShapeN) * sizeof(T) / 32,
        (realdstN - maskShapeN) * sizeof(T), 0);
    AscendC::DataCopyPad(dstTensor, srcTensor, dataCopyParams);  // 无 padParams
}
```

`src/tl_templates/ascend/common.h:215-249` — `copy_gm_to_ub`（有 padding，对比参考）

```cpp
template <typename T, uint32_t dstN, uint32_t dstM = 1>
CATLASS_DEVICE void
copy_gm_to_ub(LocalTensor<T> dstTensor, GlobalTensor<T> srcTensor, ...) {
    bool isPad = true;
    uint32_t rightPadding = 1;
    if (maskShapeN == dstN || (maskShapeN * sizeof(T)) % 32 == 0) {
        isPad = false;
        rightPadding = 0;
    }
    // ... Duplicate 填充 gap ...
    AscendC::DataCopyPadExtParams<T> padParams(isPad, 0, rightPadding, padValue);
    AscendC::DataCopyPad(dstTensor, srcTensor, dataCopyParams, padParams);  // 有 padParams
}
```

### 生成的 C++ 代码

对于 `T.reduce_max(a_ub, b_ub, dim=-1)` + `T.copy(b_ub, B)` 其中 b_ub shape=(16,1)：

```cpp
tl::ascend::reduce_max<float, 16, 128, -1>(b_ub[0], a_ub[0], tmp_ub[0], true);
tl::ascend::copy_ub_to_gm<float, 1, 16>(B[0], b_ub[0], 1, 16, 1);
//                                      ^^^^^^^^^^^^^^
//                                      srcN=1 → blockLen = 1 * 4 = 4 bytes
//                                      4 % 32 != 0 → DataCopyPad 无 padding → 数据损坏
```

### Examples 规避方式

所有 examples 中 reduce 的 `(M, 1)` keepdim 输出后续被 V-pipe 操作使用，不是直接 `copy_ub_to_gm` 输出到 GM：

- `examples/deepseek_v4/sparse_attention.py:133` — `T.reduce_max` → `T.tile.max` → `T.tile.broadcast`（V-pipe 操作链，不直接 copy_ub_to_gm）
- `examples/gqa_fwd_varlen/gqa_fwd_varlen.py:440` — 同上模式
- `examples/mha_sink_fwd_bhsd/mha_sink_fwd_bhsd.py:460` — 同上模式

5 个有精度验证的 example 全部通过（sparse_attention、gqa_fwd_varlen、mha_sink_fwd_bhsd、example_gqa_sink_fwd_varlen、split_qkv_rmsnorm_mrope），因为 reduce 结果不直接输出到 GM。

</details>
